SPD-409 Scrubbing added

lnauta opened the pull request:
The scrubbing options have been added through examples where the Actor is customised.

lnauta commented:
Note that because this MR uses functionality from #16 that #16 needs to be merged into master first.

hailihu commented:
Please merge with master and resolve conflicts.

hailihu commented:
According to user story 2:
"As a Picas user, I want the tokens that exceed a specified expected execution time to be automatically resubmitted up to X times, so that I do not need to manually reset them every time."
However, the current implementation of the maximum execution time (either through max_time or stop_function) is different. Currently, the maximum execution time does not apply to the execution time of one token but all tokens together, because the Timer is initialized before the iterator loop. Also, a token is not forced to stop if max_time is exceeded, instead it is checked only after the token is already finished if max_time is exceeded.

If I understand the user story correctly, we need to use e.g. the `stopit` library: https://theautomatic.net/2021/11/27/how-to-stop-long-running-code-in-python/

hailihu commented:
Having the scrub_count functionality in a example files (scrub-example.py and scrub-timer-example.py) creates duplicate code (i.e. more difficult to maintain) and makes it more difficult for the users as they would need to adapt more code themselves. 

I think it's better to have the scrub_count functionality in `actors.py` together with the stop functionality (`RunActor.run`). Also the scrub_count needs to be increased in case of the `handler` function is called (i.e. when Picas is killed).

hailihu commented on examples/scrub-example.py:
Having the scrub_count functionality in a example files (scrub-example.py and scrub-timer-example.py) creates duplicate code (i.e. more difficult to maintain) and makes it more difficult for the users as they would need to adapt more code themselves.

I think it's better to have the scrub_count functionality in actors.py together with the stop functionality (RunActor.run). Also the scrub_count needs to be increased in case of the handler function is called (i.e. when Picas is killed).

hailihu reviewed (changes requested):
According to user story 2:
"As a Picas user, I want the tokens that exceed a specified expected execution time to be automatically resubmitted up to X times, so that I do not need to manually reset them every time."
However, the current implementation of the maximum execution time (either through max_time or stop_function) is different. Currently, the maximum execution time does not apply to the execution time of one token but all tokens together, because the Timer is initialized before the iterator loop. Also, a token is not forced to stop if max_time is exceeded, instead it is checked only after the token is already finished if max_time is exceeded.

If I understand the user story correctly, we need to use e.g. the stopit library: https://theautomatic.net/2021/11/27/how-to-stop-long-running-code-in-python/

lnauta commented:
I added an initial attempt for the `stopit` functionality, take a look if this is what youre looking for. I have not tested it yet.

I have also added the scrubbing to the handler. As for moving it to the actor class, can you make a suggestion on the implementation?

hailihu commented:
> I added an initial attempt for the `stopit` functionality, take a look if this is what youre looking for. I have not tested it yet.
> 
> I have also added the scrubbing to the handler. As for moving it to the actor class, can you make a suggestion on the implementation?

Find suggestion here: https://github.com/sara-nl/picasclient/pull/23
You can merge if you approve into this branch, and I can continue testing and reviewing.

hailihu commented:
While testing, I ran into this issue: https://github.com/sara-nl/picasclient/pull/17. It is an already existing issue, in this case triggered an Exception when a token exceeds the `max_token_time`.  This results in "lost" tokens that are not visible in the Picas Views. Should we fix this before this PR is merged?

lnauta commented:
> Suggestion to fix this error here: #25.

Has been merged into SPD-409
